### DISCUSSION TITLE
Frontmatter Rendered on Page

### DISCUSSION THREAD

**Discussion by shroommu** [mdx-js/mdx#2349] [Q&A] [answered]
I have the following React code rendering a .mdx file:

```
import styled from "styled-components";
import { MDXProvider } from "@mdx-js/react";

import { Header1, Header2, Paragraph } from "./styles";

const Container = styled.article``;

const components = {
  h1: Header1,
  h2: Header2,
  p: Paragraph,
};

export default function BlogPost({ children }) {
  return (
    <Container>
      <MDXProvider components={components}>{children}</MDXProvider>
    </Container>
  );
}

```

And the .mdx file is as follows: 

```
---
title: "Building my Site, Part 1"
slug: "/building-my-site-part-1"
date: "09-14-2023"
excerpt: "Building a site from scratch is a fun adventure, and I'll tell you all about it in this multi-part blog series!"
---

# Hi! Here's the title!

## Here's a sub-title!

Here's some body text.

```

When I render the React component, it includes the frontmatter, which is not my expected or desired outcome: 
![image](https://github.com/mdx-js/mdx/assets/13990226/b93b244a-7dd3-4059-b00c-6e5d011904e9)

What am I doing wrong here? I've tried several different configurations of the frontmatter and different frontmatter-related plugins, to no avail. I suspect it's something with the remark plugin configuration, but I'm very unclear on how I'm supposed to configure it? I'm using webpack; here is my webpack config (relevant section):

```
{
        test: /\.mdx?$/,
        use: [
          {
            loader: "@mdx-js/loader",
            /** @type {import('@mdx-js/loader').Options} */
            options: {
              remarkPlugins: [remarkFrontmatter, { type: "yaml", marker: "-" }],
            },
          },
        ],
      },
```

**Comment by ChristianMurphy** [comment]
Welcome @shroommu!
Sorry you ran into a spot of trouble, a couple thoughts:
1. You mention configuring webpack to load MDX files, but you don't appear to use the mdx loader to `import Document from './path/file.mdx'` anywhere. Are you using a different system such as Next to load the file into `{children}`?
   * If so be sure to configure that
   * If not, how are you using the webpack loader?
2. You probably don't need to use `MDXProvider`. If you are using the webpack or next loader, the document will be passed as a component to you, and that component has a `components` prop which can be used to add custom elements.

**Reply by shroommu** [reply]
Ah good to know about passing `components` directly to the mdx component, that removes one layer of complexity 😊 

I am not calling the mdx loader directly anywhere, I've only configured it in my `webpack.config`. I'm almost positive I've configured it incorrectly, though. Here's the whole thing: 

```
const webpack = require("webpack");
const MiniCssExtractPlugin = require("mini-css-extract-plugin");

module.exports = {
  entry: {
    app: "./src/index.js",
  },
  output: {
    path: __dirname + "/dist",
    filename: "bundle.js",
    publicPath: "",
    hashFunction: "xxhash64",
  },
  plugins: [
    new MiniCssExtractPlugin({
      filename: "[name].bundle.css",
      chunkFilename: "[id].css",
    }),
    new webpack.HotModuleReplacementPlugin(),
  ],
  module: {
    rules: [
      {
        test: /\.css$/i,
        use: [
          "style-loader",
          {
            loader: "css-loader",
            options: { importLoaders: 1 },
          },
          {
            loader: "postcss-loader",
            options: {
              postcssOptions: {
                plugins: [
                  ["postcss-import"],
                  [
                    "postcss-custom-media",
                    {
                      // Options
                    },
                  ],
                ],
              },
            },
          },
        ],
      },
      // {
      //   test: /\.(png|svg|jpg|jpeg|gif)$/i,
      //   type: "asset/resource",
      // },
      {
        test: /\.(jsx|js)$/,
        exclude: /node_modules/,
        use: [
          {
            loader: "babel-loader",
            options: {
              presets: [
                [
                  "@babel/preset-env",
                  {
                    targets: "defaults",
                  },
                ],
                "@babel/preset-react",
              ],
            },
          },
        ],
      },
      {
        test: /\.mdx?$/,
        use: [
          {
            loader: "@mdx-js/loader",
            /** @type {import('@mdx-js/loader').Options} */
            options: {
              remarkPlugins: [remarkFrontmatter, { type: "yaml", marker: "-" }],
            },
          },
        ],
      },
    ],
  },
};
```

**Reply by ChristianMurphy** [reply]
For configuring webpack, see https://mdxjs.com/docs/getting-started/#webpack
You _may_ also need `babel` in your setup.

> I am not calling the mdx loader directly anywhere

Okay good to know, in that case, changing the webpack loader may not have any effect.

If not the webpack loader, my question comes back to, what is sending the `children` prop to:
```jsx
export default function BlogPost({ children }) {
```
in your example?

**Reply by shroommu** [reply]
Here's the parent component that actually imports the .mdx:

```
import Section from "../basic/Section";
import Card from "../basic/Card";
import styled from "styled-components";
import { Heading } from "../shared";
import BlogPost from "./BlogPost";

import FirstPost from "../../data/posts/9-14-2023-building-site-part-1.mdx";

const ContentContainer = styled.div`
  display: flex;
  flex-direction: column;
`;

export default function Blog() {
  return (
    <Section testId="blog-section">
      <Card zIndex={5}>
        <ContentContainer>
          <Heading>Blog</Heading>
          <BlogPost>
            <FirstPost />
          </BlogPost>
        </ContentContainer>
      </Card>
    </Section>
  );
}
```

**Reply by ChristianMurphy** [reply]
This line
```js
import FirstPost from "../../data/posts/9-14-2023-building-site-part-1.mdx";
```
would be the part that is calling the loader, based off the `test: /\.mdx?$/,` rule you have in your configuration.

In that case, and because you are also using `babel-loader` elsewhere.

```js
import remarkFrontmatter from 'remark-frontmatter';

module.exports = {
  module: {
    // …
    rules: [
      // …
      {
        test: /\.mdx?$/,
        use: [
          // `babel-loader` is optional:
          {
            loader: "babel-loader",
            options: {
              presets: [
                [
                  "@babel/preset-env",
                  {
                    targets: "defaults",
                  },
                ],
                "@babel/preset-react",
              ],
            },
          },
          {
            loader: "@mdx-js/loader",
            /** @type {import('@mdx-js/loader').Options} */
            options: {
              remarkPlugins: [remarkFrontmatter, { type: "yaml", marker: "-" }],
            },
          },
        ]
      }
    ]
  }
}
```

is probably closer to what you are looking for.

**Reply by shroommu** [reply]
Hm... Closer, maybe, but that config is giving me the same result.

I'm really at a loss for what to try next. I've read the documentation for MDX several times over, and I don't know if I'm just missing something or if what I'm trying to do is not possible with the project as I have it configured. I know CRA and MDX don't currently play nicely together, but I hesitate to dismantle everything in order to try something that might not work, either.

**Reply by ChristianMurphy** [reply]
Have you read through or tried https://mdxjs.com/guides/frontmatter/ yet? It looks like you probably want to do something with the content/yaml similar to this.

**Reply by shroommu** [reply]
I have, and I do, but it's not clear how to translate the examples on that page into something that will work with the webpack loader.

I may have to explore other options at this point. I've been pulling my hair out all day to do something that seems like it should be simple, but apparently isn't.

**Reply by shroommu** [reply, marked as answer]
WELL. Scratch that. Guess I finally pulled out the right hair.

Installed CRACO as recommended in the docs about CRA, downgraded `remark-frontmatter` to 3.0.0, and used this CRACO config file:

```
const { addAfterLoader, loaderByName } = require("@craco/craco");
const remarkFrontmatter = require("remark-frontmatter");

module.exports = {
  webpack: {
    configure: (webpackConfig) => {
      addAfterLoader(webpackConfig, loaderByName("babel-loader"), {
        test: /\.(md|mdx)$/,
        loader: require.resolve("@mdx-js/loader"),
        options: {
          providerImportSource: "@mdx-js/react",
          remarkPlugins: [remarkFrontmatter, { type: "yaml", marker: "-" }],
        },
      });

      return webpackConfig;
    },
  },
};
```

Now it works!

**Reply by ChristianMurphy** [reply]
Glad you got it working.
A fair warning, while CRA still works, it is for all intents and purposes deprecated by the React team https://github.com/reactjs/react.dev/pull/5487#issuecomment-1409720741.
If you are looking to create a single page app, Vite (which also has an MDX integration https://mdxjs.com/docs/getting-started/#vite) may be a better fit.
And if you'd like server side rendering support React has some recommendations https://react.dev/learn/start-a-new-react-project (all of which also have MDX integrations)
